catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,502 @@
+
+            if transaction_slice.nb_witnesses() != 1 {
+                unimplemented!("Multi-signature is not supported");
+            }
+
+            let witness = if let Witness::Account(witness) =
+                transaction_slice.witnesses().iter().next().unwrap()
+            {
+                witness
+            } else {
+                panic!("utxo witnesses not supported");
+            };
+
+            let (is_valid_tx, sc) = verify_original_tx(
+                spending_counter,
+                &self.old_block0_hash.into_hash(),
+                &sign_data_hash,
+                &identifier.to_inner(),
+                &witness,
+                self.range_check.clone(),
+            );
+
+            if !is_valid_tx {
+                return Err(Error::InvalidTransactionSignature {
+                    id: fragment.clone().hash().to_string(),
+                    range: self.range_check.clone(),
+                });
+            }
+
+            self.spending_counters
+                .entry(address.clone().into())
+                .or_insert_with(Vec::new)
+                .push(sc);
+
+            let wallet = self.wallets.get_mut(&address.into()).unwrap();
+
+            let vote_plan = self
+                .voteplans
+                .get(vote_cast.vote_plan())
+                .ok_or(Error::MissingVoteplanError)?;
+
+            // we still use the old block0 hash because the new ledger will still use the old one for
+            // verifications. This makes possible the usage of old VoteTally transactions without the need
+            // to be replayed.
+            let res = wallet
+                .issue_vote_cast_cert(
+                    &self.old_block0_hash,
+                    &self.fees,
+                    &vote_plan,
+                    vote_cast.proposal_index(),
+                    &choice,
+                )
+                .unwrap();
+            wallet.confirm_transaction();
+            // add to processed list so it is not processed if duplicated
+            self.processed_fragments.insert(processed_key);
+            Ok(res)
+        } else {
+            unimplemented!();
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{increment_ledger_time_up_to, recover_ledger_from_logs, voteplans_from_block0};
+    use chain_impl_mockchain::block::Block;
+    use chain_impl_mockchain::certificate::VoteTallyPayload;
+    use chain_impl_mockchain::vote::Weight;
+    use chain_ser::deser::Deserialize;
+    use jormungandr_lib::interfaces::{
+        load_persistent_fragments_logs_from_folder_path, Block0Configuration,
+    };
+    use jormungandr_testing_utils::wallet::Wallet;
+    use std::io::BufReader;
+    use std::path::PathBuf;
+
+    fn read_block0(path: PathBuf) -> std::io::Result<Block> {
+        let reader = std::fs::File::open(path)?;
+        Ok(Block::deserialize(BufReader::new(reader)).unwrap())
+    }
+
+    #[test]
+    fn test_vote_flow() -> std::io::Result<()> {
```

Comment:
Is there a description for the testing data? I think we are interested in whether it has unordered fragments.

Reply:
There is another PR incoming with better testing which actually tests unordered fragments.